execution/shuffle.go

```diff
@@ -0,0 +1,449 @@
+	return &Shuffle{
+		outputPartitionCount: outputPartitionCount,
+		strategyPrototype:    strategyPrototype,
+		sources:              sources,
+	}
+}
+
+func (s *Shuffle) Get(ctx context.Context, variables octosql.Variables, streamID *StreamID) (RecordStream, *ExecutionOutput, error) {
+	pipelineMetadata := ctx.Value(pipelineMetadataContextKey{}).(PipelineMetadata)
+
+	receiver := NewShuffleReceiver(streamID, pipelineMetadata.NextShuffleID, len(s.sources), pipelineMetadata.Partition)
+	execOutput := NewExecutionOutput(receiver, map[string]ShuffleData{
+		pipelineMetadata.NextShuffleID.AsMapKey(): {
+			Shuffle:   s,
+			Variables: variables,
+			ShuffleID: pipelineMetadata.NextShuffleID,
+		},
+	})
+
+	return receiver, execOutput, nil
+}
+
+func (s *Shuffle) StartSources(ctx context.Context, stateStorage storage.Storage, tx storage.StateTransaction, shuffleID *ShuffleID, variables octosql.Variables) (map[string]ShuffleData, error) {
+	// Create an ID for the possible _even next_ shuffle.
+	nextShuffleIDRaw, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString("next_shuffle_id"))
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get next shuffle id")
+	}
+	nextShuffleID := (*ShuffleID)(nextShuffleIDRaw)
+
+	nextShuffles := make(map[string]ShuffleData)
+
+	for partition, node := range s.sources {
+		sourceStreamID, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString(fmt.Sprintf("shuffle_input_%d", partition)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream ID for shuffle with ID %s", shuffleID.Id)
+		}
+
+		newPipelineMetadata := PipelineMetadata{
+			NextShuffleID: nextShuffleID,
+			Partition:     partition,
+		}
+
+		// Get the shuffle source stream.
+		rs, execOutput, err := node.Get(context.WithValue(ctx, pipelineMetadataContextKey{}, newPipelineMetadata), variables, sourceStreamID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream for shuffle with ID %s", shuffleID.Id)
+		}
+
+		// Add deduplicated next shuffle information.
+		for id, data := range execOutput.NextShuffles {
+			nextShuffles[id] = data
+		}
+
+		senderStreamID, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString(fmt.Sprintf("shuffle_sender_%d", partition)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream ID for shuffle with ID %s", shuffleID.Id)
+		}
+
+		strategy, err := s.strategyPrototype.Get(ctx, variables)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get strategy for shuffle with ID %s", shuffleID.Id)
+		}
+
+		// Start the shuffle sender.
+		sender := NewShuffleSender(senderStreamID, shuffleID, strategy, s.outputPartitionCount, partition)
+
+		engine := NewPullEngine(sender, stateStorage, rs, nil, execOutput.WatermarkSource, false)
+
+		go engine.Run(ctx)
+	}
+
+	return nextShuffles, nil
+}
+
+type ShuffleReceiver struct {
```

Comment:
Please, if you introduce a structure which purpose isn't totally obvious, write a small comment explaining what it does like "ShuffleReceiver reads data from underlying shuffles and merges it" (I know it's prolly not what it does, but you get the point)

Reply:
It was obvious to me :P but I'll add comments